zerver/tests/test_auth_backends.py

```diff
@@ -1661,6 +1661,203 @@ def test_social_auth_saml_require_limit_to_subdomains(self) -> None:
+    LOGIN_URL = "/accounts/login/social/apple"
+    SIGNUP_URL = "/accounts/register/social/apple"
+    AUTHORIZATION_URL = "https://appleid.apple.com/auth/authorize"
+    ACCESS_TOKEN_URL = "https://appleid.apple.com/auth/token"
+    AUTH_FINISH_URL = "/complete/apple/"
+    CONFIG_ERROR_URL = "/config-error/apple"
+
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         desktop_flow_otp: Optional[str]=None,
+                         is_signup: bool=False,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         **extra_data: Any) -> HttpResponse:
+        url, headers = self.prepare_login_url_and_headers(
+            subdomain, mobile_flow_otp, desktop_flow_otp, is_signup, next,
+            multiuse_object_key
+        )
+
+        result = self.client_get(url, **headers)
+
+        expected_result_url_prefix = 'http://testserver/login/%s/' % (self.backend.name,)
+        if settings.SOCIAL_AUTH_SUBDOMAIN is not None:
+            expected_result_url_prefix = ('http://%s.testserver/login/%s/' %
+                                          (settings.SOCIAL_AUTH_SUBDOMAIN, self.backend.name,))
+
+        if result.status_code != 302 or not result.url.startswith(expected_result_url_prefix):
+            return result
+
+        result = self.client_get(result.url, **headers)
+        self.assertEqual(result.status_code, 302)
+        assert self.AUTHORIZATION_URL in result.url
+
+        self.client.cookies = result.cookies
+
+        # id_token in a JSON Web Token that contains the user’s identity information.
+        token_data_dict = {
+            'access_token': 'foobar',
+            'expires_in': time.time() + 60*5,
+            'id_token': self.generate_id_token(account_data_dict),
+            'token_type': 'bearer'
+        }
+
+        parsed_url = urllib.parse.urlparse(result.url)
+        state = urllib.parse.parse_qs(parsed_url.query)['state']
+
+        with responses.RequestsMock(assert_all_requests_are_fired=False) as requests_mock:
+            requests_mock.add(
+                requests_mock.POST,
+                self.ACCESS_TOKEN_URL,
+                match_querystring=False,
+                status=200,
+                body=json.dumps(token_data_dict))
+
+            # The below mocked function usually gets jwk's from
+            # https://appleid.apple.com/auth/keys. `decode_token`
+            # finds the right jwk with help of `kid` obtained with
+            # `jwt.get_unverified_header`. Here, because we generate
+            # `id_token` we have the right public key to decode it.
+            # This mock return's that public key in jwk format.
+            with mock.patch('social_core.backends.apple.AppleIdAuth.get_apple_jwk',
+                            return_value=settings.APPLE_JWK):
```

Comment:
Could this be replaced with a sample RequestsMock of the apple url to return the JWK as http response?

Reply:
yes, replaced it.